src/Abstractions/TaskName.cs

```diff
@@ -8,127 +8,162 @@ namespace Microsoft.DurableTask;
+  }
 
-    /// <summary>
-    /// Gets a value indicating whether to <see cref="TaskName"/> objects
-    /// are equal using value semantics.
-    /// </summary>
-    /// <param name="other">The other object to compare to.</param>
-    /// <returns><c>true</c> if the two objects are equal using value semantics; otherwise <c>false</c>.</returns>
-    public bool Equals(TaskName other)
+  /// <summary>
+  /// Gets a value indicating whether to <see cref="TaskName"/> objects
+  /// are equal using value semantics.
+  /// </summary>
+  /// <param name="obj">The other object to compare to.</param>
+  /// <returns><c>true</c> if the two objects are equal using value semantics; otherwise <c>false</c>.</returns>
+  public override bool Equals(object? obj)
+  {
+    if (obj is not TaskName other)
     {
-        return string.Equals(this.Name, other.Name, StringComparison.OrdinalIgnoreCase);
+      return false;
     }
 
-    /// <summary>
-    /// Gets a value indicating whether to <see cref="TaskName"/> objects
-    /// are equal using value semantics.
-    /// </summary>
-    /// <param name="obj">The other object to compare to.</param>
-    /// <returns><c>true</c> if the two objects are equal using value semantics; otherwise <c>false</c>.</returns>
-    public override bool Equals(object? obj)
-    {
-        if (obj is not TaskName other)
-        {
-            return false;
-        }
+    return this.Equals(other);
+  }
 
-        return this.Equals(other);
-    }
+  /// <summary>
+  /// Calculates a hash code value for the current <see cref="TaskName"/> instance.
+  /// </summary>
+  /// <returns>A 32-bit hash code value.</returns>
+  public override int GetHashCode()
+  {
+    return StringComparer.OrdinalIgnoreCase.GetHashCode(this.Name);
+  }
 
-    /// <summary>
-    /// Calculates a hash code value for the current <see cref="TaskName"/> instance.
-    /// </summary>
-    /// <returns>A 32-bit hash code value.</returns>
-    public override int GetHashCode()
+  /// <summary>
+  /// Gets the string value of the current <see cref="TaskName"/> instance.
+  /// </summary>
+  /// <returns>The name and optional version of the current <see cref="TaskName"/> instance.</returns>
+  public override string ToString()
+  {
+    if (string.IsNullOrEmpty(this.Version))
     {
-        return StringComparer.OrdinalIgnoreCase.GetHashCode(this.Name);
+      return this.Name;
     }
-
-    /// <summary>
-    /// Gets the string value of the current <see cref="TaskName"/> instance.
-    /// </summary>
-    /// <returns>The name and optional version of the current <see cref="TaskName"/> instance.</returns>
-    public override string ToString()
+    else
     {
-        if (string.IsNullOrEmpty(this.Version))
-        {
-            return this.Name;
-        }
-        else
-        {
-            return this.Name + ":" + this.Version;
-        }
+      return this.Name + ":" + this.Version;
```

Comment:
We should consider adding FromString semantics. But I wonder if that will be a breaking change? We have not restricted what characters can be used in a task name before this.

This change is going to introduce a confusing behavior:

``` CSharp
TaskName name1 = new("MyTask", "2"); // Name = "MyTask", Version = "2'
TaskName name2 = name1.ToString(); // implicit operator convers. Name = "MyTask:2", Version = ""
```

Reply:
Should the implicit operator be updated to parse out the version?

I'm not super concerned about the breaking change since 99+% of .NET users should be relying on the function name, which already can't have special characters. However, we can take a look at Kusto data to confirm.

Reply:
Yes the implicit operator would be changed to call `FromString`.